### PR TITLE
Move RecoverVolumeExpansionFailure feature to beta

### LINKED ISSUE #115294: NodeExpandVolume gets called, even if CSI plugin returned NodeExpansionRequired=false in ControllerExpandVolumeResponse
### What happened?

In case of online volume expansion of raw block volumes, some of the CSI plugins may not need file system resize or any related handling and thus NodeExpandVolume API call can be skipped for the same. In order to do this, plugin must set NodeExpansionRequired to false, in ControllerExpandVolumeResponse.
However, it is observed that kubelet does call NodeExpandVolume API, after successful resize done by CSI plugin in ControllerExpandVolume API call with NodeExpansionRequired=false in response.
Even though CSI plugin can skip handling in NodeExpandVolume API as per its requirement, is it expected to get this API call from kubelet when NodeExpansionRequired set to false by plugin in ControllerExpandVolumeResponse?

### What did you expect to happen?

During CSI volume online expansion, if CSI plugin returns NodeExpansionRequired=false in ControllerExpandVolumeResponse, to indicate no FS resize handling needed, then kubelet should not call NodeExpandVolume API for such volume(s).

### How can we reproduce it (as minimally and precisely as possible)?

The issue is seen with vSphere CSI driver, while expanding vSphere raw-block volume (volumeMode: Block), already in use by one pod(raw-block-pod). However, the issue may not be specific to vSphere CSI driver.

```
# kubectl describe pvc raw-block-pvc
Name:          raw-block-pvc
Namespace:     default
StorageClass:  example-raw-block-sc
Status:        Bound
Volume:        pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d
Labels:        <none>
Annotations:   [pv.kubernetes.io/bind-completed](http://pv.kubernetes.io/bind-completed): yes
               [pv.kubernetes.io/bound-by-controller](http://pv.kubernetes.io/bound-by-controller): yes
               [volume.beta.kubernetes.io/storage-provisioner](http://volume.beta.kubernetes.io/storage-provisioner): [csi.vsphere.vmware.com](http://csi.vsphere.vmware.com/)
               [volume.kubernetes.io/storage-provisioner](http://volume.kubernetes.io/storage-provisioner): [csi.vsphere.vmware.com](http://csi.vsphere.vmware.com/)
Finalizers:    [[kubernetes.io/pvc-protection](http://kubernetes.io/pvc-protection)]
Capacity:      4Gi
Access Modes:  RWO
VolumeMode:    Block
Used By:       raw-block-pod
Events:
  Type     Reason                      Age                From                                                                                                 Message
  ----     ------                      ----               ----                                                                                                 -------
  Normal   Provisioning                17m                [csi.vsphere.vmware.com](http://csi.vsphere.vmware.com/)_vsphere-csi-controller-66f5d46f77-cbvxr_be3aefaf-00a8-4a98-8e40-a42ff01c997a  External provisioner is provisioning volume for claim "default/raw-block-pvc"
  Normal   ExternalProvisioning        17m (x2 over 17m)  persistentvolume-controller                                                                          waiting for a volume to be created, either by external provisioner "[csi.vsphere.vmware.com](http://csi.vsphere.vmware.com/)" or manually created by system administrator
  Normal   ProvisioningSucceeded       17m                [csi.vsphere.vmware.com](http://csi.vsphere.vmware.com/)_vsphere-csi-controller-66f5d46f77-cbvxr_be3aefaf-00a8-4a98-8e40-a42ff01c997a  Successfully provisioned volume pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d
  Warning  ExternalExpanding           85s (x2 over 16m)  volume_expand                                                                                        Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
  Normal   Resizing                    85s (x2 over 16m)  external-resizer [csi.vsphere.vmware.com](http://csi.vsphere.vmware.com/)                                                              External resizer is resizing volume pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d
  Normal   VolumeResizeSuccessful      84s (x2 over 16m)  external-resizer [csi.vsphere.vmware.com](http://csi.vsphere.vmware.com/)                                                              Resize volume succeeded
  Normal   FileSystemResizeSuccessful  59s                kubelet                                                                                              MountVolume.NodeExpandVolume succeeded for volume "pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d" k8s-node-803-1673532986

```

### Anything else we need to know?

Kubelet logs on node where raw-block-pod was running:

`Jan 23 05:46:53 k8s-node-803-1673532986 kubelet[389705]: I0123 05:46:53.362905  389705 operation_generator.go:2217] "MountVolume.NodeExpandVolume succeeded for volume \"pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d\" (UniqueName: \"kubernetes.io/csi/csi.vsphere.vmware.com^81b5bb35-d858-4f43-b8a0-8dcafa967934\") pod \"raw-block-pod\" (UID: \"4ac3bfad-cf20-425d-a22c-6533d7a91a41\") k8s-node-803-1673532986" pod="default/raw-block-pod"
`

CSI resizer logs:

```
I0123 05:46:43.696297       1 controller.go:295] Started PVC processing "default/raw-block-pvc"
I0123 05:46:43.713184       1 event.go:285] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"raw-block-pvc", UID:"ad41ea4d-725b-42d7-acb5-9aea7679d17d", APIVersion:"v1", ResourceVersion:"3288118", FieldPath:""}): type: 'Normal' reason: 'Resizing' External resizer is resizing volume pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d
I0123 05:46:44.857962       1 controller.go:483] Resize volume succeeded for volume "pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d", start to update PV's capacity
I0123 05:46:44.858044       1 controller.go:590] Resize volume succeeded for volume "pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d", start to update PV's capacity
I0123 05:46:44.888197       1 controller.go:489] Update capacity of PV "pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d" to 4Gi succeeded
I0123 05:46:44.909795       1 controller.go:546] Resize PVC "default/raw-block-pvc" finished
I0123 05:46:44.909862       1 controller.go:295] Started PVC processing "default/raw-block-pvc"
I0123 05:46:44.909873       1 controller.go:343] No need to resize PVC "default/raw-block-pvc"
I0123 05:46:44.909948       1 event.go:285] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"raw-block-pvc", UID:"ad41ea4d-725b-42d7-acb5-9aea7679d17d", APIVersion:"v1", ResourceVersion:"3288118", FieldPath:""}): type: 'Normal' reason: 'VolumeResizeSuccessful' Resize volume succeeded
```

vSphere CSI controller logs:

```
2023-01-23T05:46:43.714Z        INFO    vanilla/controller.go:2362      ControllerExpandVolume: called with args {VolumeId:81b5bb35-d858-4f43-b8a0-8dcafa967934 CapacityRange:required_bytes:4294967296  Secrets:map[] VolumeCapability:block:<> access_mode:<mode:SINGLE_NODE_MULTI_WRITER >  XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}     {"TraceId": "cf237bb1-cbfc-4c9d-91e1-d22316d38532"}
...
2023-01-23T05:46:43.741Z        DEBUG   common/vsphereutil.go:629       vSphere CSI driver expanding volume "81b5bb35-d858-4f43-b8a0-8dcafa967934" to new size 4096 Mb.
...
2023-01-23T05:46:44.841Z        INFO    volume/manager.go:1594  ExpandVolume: volumeID: "81b5bb35-d858-4f43-b8a0-8dcafa967934", opId: "44860a6a"        {"TraceId": "cf237bb1-cbfc-4c9d-91e1-d22316d38532"}
2023-01-23T05:46:44.841Z        INFO    volume/manager.go:1634  ExpandVolume: Volume expanded successfully to size 4096. volumeID: "81b5bb35-d858-4f43-b8a0-8dcafa967934", opId: "44860a6a"     {"TraceId": "cf237bb1-cbfc-4c9d-91e1-d22316d38532"
...
2023-01-23T05:46:44.857Z        DEBUG   volume/manager.go:1388  internalExpandVolume: returns fault "" for volume "81b5bb35-d858-4f43-b8a0-8dcafa967934"        {"TraceId": "cf237bb1-cbfc-4c9d-91e1-d22316d38532"}
2023-01-23T05:46:44.857Z        INFO    common/vsphereutil.go:663       Successfully expanded volume for volumeID "81b5bb35-d858-4f43-b8a0-8dcafa967934" to new size 4096 Mb.   {"TraceId": "cf237bb1-cbfc-4c9d-91e1-d22316d38532"}
2023-01-23T05:46:44.857Z        DEBUG   vanilla/controller.go:2498      ControllerExpandVolumeInternal: returns '4096' as capacity and false as NodeExpansionRequired       {"TraceId": "a8e59b5e-a6f7-4a99-8fbb-030e6f199575"}
2023-01-23T05:46:44.857Z        INFO    vanilla/controller.go:2460      Volume "81b5bb35-d858-4f43-b8a0-8dcafa967934" expanded successfully.    {"TraceId": "cf237bb1-cbfc-4c9d-91e1-d22316d38532"}
```

vSphere CSI node plugin logs:

```
2023-01-23T05:46:53.274Z        INFO    service/node.go:480     NodeExpandVolume: called with args {VolumeId:81b5bb35-d858-4f43-b8a0-8dcafa967934 VolumePath:/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d/4ac3bfad-cf20-425d-a22c-6533d7a91a41 CapacityRange:required_bytes:4294967296  StagingTargetPath:/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d VolumeCapability:block:<> access_mode:<mode:SINGLE_NODE_MULTI_WRITER >  Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}     {"TraceId": "2f6048a9-ac00-443f-bb20-8b04a8d2096c"}
2023-01-23T05:46:53.278Z        INFO    osutils/linux_os_utils.go:627   check path exits /dev/sdb       {"TraceId": "2f6048a9-ac00-443f-bb20-8b04a8d2096c"}
2023-01-23T05:46:53.278Z        DEBUG   service/node.go:516     NodeExpandVolume: staging target path /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d/4ac3bfad-cf20-425d-a22c-6533d7a91a41, getDevFromMount {FullPath:/dev/sdb Name:sdb RealDev:/dev/sdb}     {"TraceId": "2f6048a9-ac00-443f-bb20-8b04a8d2096c"}
2023-01-23T05:46:53.362Z        DEBUG   service/node.go:543     NodeExpandVolume: Resized filesystem with devicePath /dev/sdb volumePath /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d/4ac3bfad-cf20-425d-a22c-6533d7a91a41 {"TraceId": "2f6048a9-ac00-443f-bb20-8b04a8d2096c"}
2023-01-23T05:46:53.362Z        INFO    service/node.go:545     NodeExpandVolume: expanded volume successfully. devicePath /dev/sdb volumePath /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-ad41ea4d-725b-42d7-acb5-9aea7679d17d/4ac3bfad-cf20-425d-a22c-6533d7a91a41 size 4294967296   {"TraceId": "2f6048a9-ac00-443f-bb20-8b04a8d2096c"}

```
### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:33:49Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:27:13Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
vSphere Cloud Provider Interface
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by gnufied**
Also fixes https://github.com/kubernetes/kubernetes/issues/115294

xref https://github.com/kubernetes/enhancements/issues/1790

```release-note
Promoted `RecoverVolumeExpansionFailure` feature gate to beta.
```

### CODE REVIEW

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -488,6 +488,7 @@ const (
 
 	// owner: @gnufied
 	// kep: https://kep.k8s.io/1790
+	// beta - v1.32
```
Comment (Jefftree) [inline pkg/features/kube_features.go:491]: nit: comment is unnecessary now

File: pkg/volume/util/operationexecutor/node_expander.go [hunk 2]
```diff
@@ -118,6 +120,20 @@ func (ne *NodeExpander) runPreCheck() bool {
...
+		return false
+	}
+
+	for _, mode := range accessModes {
+		if mode == v1.ReadWriteMany {
+			return true
+		}
+	}
+	return false
+}
```
Comment (jsafrane) [inline pkg/volume/util/operationexecutor/node_expander.go:135]: use ContainsAccessMode:
https://github.com/kubernetes/kubernetes/blob/5f594f42159f2ee84f805d382638f77dd5967177/staging/src/k8s.io/kubectl/pkg/util/storage/storage.go#L78-L78

File: pkg/volume/util/operationexecutor/node_expander_test.go [hunk 3]
```diff
@@ -110,10 +110,21 @@ func TestNodeExpander(t *testing.T) {
...
 			pv:   getTestPV("test-vol0", "2G"),
 
+			expectedResizeStatus:     "",
+			expectResizeCall:         false,
+			assumeResizeOpAsFinished: true,
+			expectFinalErrors:        false,
+			expectedStatusSize:       resource.MustParse("2G"),
+		},
+		{
+			name: "RWOP volumes, pv.spec.cap = pvc.status.cap, resizeStatus='', desiredSize > actualSize",
```
Comment (jsafrane) [inline pkg/volume/util/operationexecutor/node_expander_test.go:124]:
```suggestion
			name: "RWX volumes, pv.spec.cap = pvc.status.cap, resizeStatus='', desiredSize > actualSize",
```

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Move RecoverVolumeExpansionFailure feature to beta
+Promoted `RecoverVolumeExpansionFailure` feature gate to beta.
```

**Comment by jsafrane** [comment]
/retest
> Verify that scheduling of a pod that uses PVC that is being deleted fails and the pod becomes Unschedulable

**Comment by tengqm** [comment]
Are we updating the website for this feature promotion?

**Pull request merged**